src/jvm/main/org/jetbrains/kotlinx/lincheck/TraceDebugger.kt

```diff
@@ -10,4 +10,85 @@
+import org.jetbrains.kotlinx.lincheck.verifier.Verifier
+import java.lang.reflect.Method
+import java.lang.reflect.Modifier
+import kotlin.reflect.jvm.kotlinFunction
+
 val isInTraceDebuggerMode = System.getProperty("lincheck.traceDebuggerMode", "false").toBoolean()
+
+internal object TraceDebuggerInjections {
+    val classUnderTraceDebugging: String? = System.getProperty("traceDebugger.className")
+    val methodUnderTraceDebugging: String? = System.getProperty("traceDebugger.methodName")
+    @JvmStatic
+    var firstRun = true
+
+    @JvmStatic
+    fun runWithLincheck(testClassName: String, testMethodName: String) {
+        firstRun = false
+
+        val testClass = Class.forName(testClassName)
+        val testMethod = testClass.getMethod(testMethodName)!!
+
+        val isStaticMethod = Modifier.isStatic(testMethod.modifiers)
+        val instanceClass = if (isStaticMethod) TraceDebuggerStaticMethodWrapper::class.java else testClass
+
+        val scenario = scenario {
+            parallel {
+                thread {
+                    // TODO: current implementation does not support `testMethod`
+                    //  that accepts arguments (e.g. java main function)
+                    if (isStaticMethod) {
+                        actor(
+                            TraceDebuggerStaticMethodWrapper::callStaticMethod,
+                            testClass, testMethod
+                        )
+                    } else {
+                        actor(testMethod.kotlinFunction!!)
+                    }
+                }
+            }
+        }
+
+        val lincheckOptions = ModelCheckingOptions()
+            .iterations(0)
+            .invocationsPerIteration(1)
+            .addCustomScenario(scenario)
+            .addGuarantee(forClasses(TraceDebuggerInjections::class).allMethods().ignore())
+            .verifier(FailingVerifier::class.java)
+            .logLevel(LoggingLevel.OFF)
+            .invocationTimeout(5 * 60 * 1000) // 5 mins
+
+        @Suppress("INVISIBLE_MEMBER", "INVISIBLE_REFERENCE")
+        val failure = lincheckOptions.checkImpl(instanceClass)
+
+        val result = failure!!.results.threadsResults[0][0]
+        if (result is ExceptionResult) throw result.throwable
+        val trace = constructTraceForPlugin(failure, failure.trace!!)
+        println(trace.joinToString("\n"))
+        // Otherwise, we just finish. For simplicity, the function always returns nothing.
+    }
+
+    @JvmStatic
+    fun isFirstRun(): Boolean {
+        return firstRun
+    }
+
+    class FailingVerifier(@Suppress("UNUSED_PARAMETER") sequentialSpecification: Class<*>) : Verifier {
+        override fun verifyResults(scenario: ExecutionScenario?, results: ExecutionResult?) = false
+    }
+}
+
+internal class TraceDebuggerStaticMethodWrapper {
+    fun callStaticMethod(clazz: Class<*>, method: Method) {
+        ensureClassHierarchyIsTransformed(clazz.name)
```

Comment:
In the trace debugger mode, let's transform ALL classes, never calling `ensureClassHierarchyIsTransformed(..)` and the reflection calls that preceed it.

Reply:
Too much classes are transformed I guess. I wasn't able to wait until the trace would be collected and shown by the plugin, it was endlessly spinning with IDEA background task "Collecting trace..."

Reply:
Oh, wow. OK, let's postpone this change then.